src/openrct2/core/FileIndex.hpp

```diff
@@ -178,32 +179,88 @@ class FileIndex
+                    size_t rangeEnd,
+                    std::vector<TItem>& items,
+                    std::mutex& printLock) const
     {
-        std::vector<TItem> items;
-        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
-
-        auto startTime = std::chrono::high_resolution_clock::now();
-        // Start at 1, so that we can reach 100% completion status
-        size_t i = 1;
-        for (auto filePath : scanResult.Files)
+        items.reserve(rangeEnd - rangeStart);
+        for (size_t i = rangeStart; i < rangeEnd; i++)
         {
-            Console::WriteFormat("File %5d of %d, done %3d%%\r", i, scanResult.Files.size(), i * 100 / scanResult.Files.size());
-            i++;
-            log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            const auto& filePath = scanResult.Files.at(i);
+
+            if (_log_levels[DIAGNOSTIC_LEVEL_VERBOSE])
+            {
+                std::unique_lock<std::mutex> lock(printLock);
+                log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            }
+
             auto item = Create(filePath);
             if (std::get<0>(item))
             {
                 items.push_back(std::get<1>(item));
             }
         }
+    }
+
+    std::vector<TItem> Build(const ScanResult &scanResult) const
+    {
+        std::vector<TItem> allItems;
+        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
+
+        auto startTime = std::chrono::high_resolution_clock::now();
+
+        const size_t totalCount = scanResult.Files.size();
+        if (totalCount > 0)
+        {
+            JobPool jobPool;
+            std::mutex printLock; // For verbose prints.
 
-        WriteIndexFile(scanResult.Stats, items);
+            std::vector<std::vector<TItem>> containers;
+
+            size_t stepSize = std::thread::hardware_concurrency();
+            size_t numTasks = totalCount / stepSize;
+            size_t taskGroup = 0;
+            numTasks += (totalCount % stepSize == 0 ? 0 : 1);
+
+            containers.resize(numTasks);
+
+            for (size_t rangeStart = 0; rangeStart < totalCount; rangeStart += stepSize, taskGroup++)
+            {
+                if (rangeStart + stepSize > totalCount)
+                    stepSize = totalCount - rangeStart;
+
+                auto& items = containers[taskGroup];
+
+                jobPool.addTask(std::bind(&FileIndex<TItem>::BuildRange,
+                    this,
+                    std::cref(scanResult),
+                    rangeStart,
+                    rangeStart + stepSize,
+                    std::ref(items),
+                    std::ref(printLock)));
+            }
+
+            jobPool.join([&]()
+            {
+                size_t completed = totalCount - jobPool.countPending();
+                Console::WriteFormat("File %5d of %d, done %3d%%\r", completed, totalCount, completed / totalCount);
```

Comment:
`completed * 100 / totalCount` to get percentages, otherwise it'll be 0 until it's finished.

Reply:
Whops.